server/src/main/java/org/opensearch/index/store/CompositeDirectory.java

```diff
@@ -0,0 +1,443 @@
+     * Currently deleting the local files here, as once uploaded to Remote, local files become eligible for eviction from FileCache
+     * @param files : recent files which have been successfully uploaded to Remote Store
+     * @throws IOException in case of I/O error
+     */
+    public void afterSyncToRemote(Collection<String> files) throws IOException {
+        logger.trace("afterSyncToRemote called for {}", files);
+        if (remoteDirectory == null) {
+            logger.trace("afterSyncToRemote called even though remote directory is not set");
+            return;
+        }
+        for (String fileName : files) {
+            writeLock.lock();
+            try {
+                /**
+                 * TODO - Unpin the files here  from FileCache so that they become eligible for eviction, once pinning/unpinning support is added in FileCache
+                 * Uncomment the below commented line(to remove the file from cache once uploaded) to test block based functionality
+                 */
+                logger.trace("File {} uploaded to Remote Store and now can be eligible for eviction in FileCache", fileName);
+                fileCache.remove(localDirectory.getDirectory().resolve(fileName));
+            } finally {
+                writeLock.unlock();
+            }
+        }
+    }
+
+    private boolean isTempFile(String name) {
+        return name.endsWith(".tmp");
+    }
+
+    /**
+     * Return the list of files present in Remote
+     */
+    private String[] getRemoteFiles() throws IOException {
+        String[] remoteFiles;
+        try {
+            remoteFiles = remoteDirectory.listAll();
+        } catch (NullPointerException e) {
+            /*
+             * There are two scenarios where the listAll() call on remote directory returns NullPointerException:
+             * - When remote directory is not set
+             * - When init() of remote directory has not yet been called
+             *
+             * Returning an empty list in these scenarios
+             */
+            remoteFiles = new String[0];
+        }
+        return remoteFiles;
+    }
+
+    private void cacheFile(String name) throws IOException {
+        Path filePath = localDirectory.getDirectory().resolve(name);
+        fileCache.put(filePath, new WrappedCachedIndexInput(localDirectory.openInput(name, IOContext.READ)));
+    }
+
+    private class CacheableIndexOutput extends FilterIndexOutput {
+
+        String fileName;
+
+        public CacheableIndexOutput(IndexOutput out, String fileName) {
+            super("CacheableIndexOutput for file : " + fileName, out);
+            this.fileName = fileName;
+        }
+
+        @Override
+        public void close() throws IOException {
+            super.close();
+            cacheFile(fileName);
+        }
+    }
+
+    private class RemoteDirectoryBlobStreamReader implements TransferManager.BlobStreamReader {
```

Comment:
We modified the Transfer manager to take a reader function as input.
The reader function for remote directory requires a context that needs to be passed, which is provided to us in the openInput function of our directory as an input. Since different function calls can have different context values passed as argument, we are storing using this class to store the context and have a single Transfer Manager which is passed to the OnDemandSnapshotBlockIndexInput. Another alternative to this class I could think of could be creating a new instance of the TransferManager class every time in the openInput function (this one's seems better imo)